packages/grafana-ui/src/components/Table/TableNG/types.ts

```diff
@@ -261,3 +262,30 @@ export interface ScrollPosition {
   x: number;
   y: number;
 }
+
+export interface TypographyCtx {
+  ctx: CanvasRenderingContext2D;
+  font: string;
+  avgCharWidth: number;
+  count: Count;
+  estimateLines: LineCounter;
+  wrappedCount: LineCounter;
+}
+
+export type LineCounter = (value: unknown, width: number) => number;
+export interface LineCounterEntry {
+  /**
+   * given a values and the available width, returns the line count for that value
+   */
+  counter: LineCounter;
+  /**
+   * if getting an accurate line count is expensive, you can provide an estimate method
+   * which will be used when looping over the row. the counter method will only be invoked
+   * for the cell which is the maximum line count for the row.
+   */
+  estimate?: LineCounter;
```

Comment:
here's the trick to making the heuristic work again: an optional `estimate` property on `LineCounterEntry`. if it exists, we call that _instead_ of counter and we only call `counter` on the max `value/width`.